File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    bool lo = false;
public:
    void addi(TreeNode* root ,int  current , int targetSum){
        if (!root) return ;
        int target = current + root->val;
        if(!root->left && !root->right && target == targetSum){
            lo = true;
        }
        addi(root->left , target , targetSum);
        addi(root->right , target , targetSum);
    }

    bool hasPathSum(TreeNode* root, int targetSum) {
        if (!root) return false;
        if(!root->left && !root->right && root->val == targetSum) return true;
        addi(root->left ,root->val,  targetSum);
        addi(root->right,root->val,targetSum);
        return lo;
    }
};

File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        def traverse(root , su):
            nonlocal targetSum
            if not root : return 
            if su + root.val == targetSum and not root.left and not root.right: return True
            return traverse(root.left , su +root.val) or traverse(root.right , su + root.val)
            
        if traverse(root , 0): return True
        else: return False